### PR TITLE
Tweak propagation of individual transactions in network

### PR DESCRIPTION
**Pull request by NikVolf**

### CODE REVIEW

**Comment by bkchr** [comment]
As already said in Riot, I'm against this solution. Why not create a proper implementation directly?

**Comment by NikVolf** [comment]
> As already said in Riot, I'm against this solution. Why not create a proper implementation directly?

Why do we need to complicate protocol if this will be enough?

It is used in Ethereum and other networks successfully

**Comment by NikVolf** [comment]
Also, this is complementary

If we choose to complicate protocol, we still want to flood sqrt(peer_count) transactions immediately to reduce latency

And announce new transactions hashes to everyone.

Like in this proposal:
https://github.com/ethereum/EIPs/issues/2465

**Comment by bkchr** [comment]
> To cater for degenerate topologies, transactions cannot be broadcast logarithmically, rather they need to be transferred linearly to all peers. With N peers, each node will transfer the same transaction N times (counting both directions), whereas 1 would be enough in a perfect world. This is a significant waste.

It says that they don't use logarithmically propagation for transactions.

**Comment by NikVolf** [comment]
> It says that they don't use logarithmically propagation for transactions.

They do with this eip (it's already live)

But yeah, probably this alone is not enough due to topology issues, I'll add announce/retrival here as well

File: client/network/src/protocol.rs [hunk 1]
```diff
@@ -1192,6 +1203,8 @@ impl<B: BlockT, H: ExHashT> Protocol<B, H> {
 	}
 
 	/// Propagate one transaction.
+	///
+	/// Use this when new valid transsaction with `hash` becomes known.
```
Comment (tomusdrw) [inline client/network/src/protocol.rs:1207]:
```suggestion
	/// Use this when new valid transaction with `hash` becomes known.
```

**Comment by NikVolf** [comment]
@tomusdrw we already have timer-based propagation on timer

also see https://github.com/paritytech/substrate/issues/6604

**Comment by tomusdrw** [comment]
> @tomusdrw we already have timer-based propagation for every peer on timer (once in 3s)

That's why I said I think it should be the other way around. The timer-based propagation (`propagate_transactions/0`) should be `sqrt` while newly imported transactions can be sent to everyone (since the packets are small) - we might introduce some batching though.

**Comment by NikVolf** [comment]
> > @tomusdrw we already have timer-based propagation for every peer on timer (once in 3s)
> 
> That's why I said I think it should be the other way around. The timer-based propagation (`propagate_transactions/0`) should be `sqrt` while newly imported transactions can be sent to everyone (since the packets are small) - we might introduce some batching though.

ah, right

@tomaka just mentioned that in new (current) protocol messages are batched automatically, so if you dispatch several messages for particular peer, they'll go batched on wire

**Comment by NikVolf** [comment]
>  The timer-based propagation

@tomusdrw I think with #6604 resolved we can just drop timer-based propagation

File: client/network/src/protocol.rs [hunk 2]
```diff
@@ -1202,21 +1215,45 @@ impl<B: BlockT, H: ExHashT> Protocol<B, H> {
...
+					.iter()
+					.filter(|(_, peer)| peer.info.roles.is_full())
+					.map(|(id, _)| id.clone())
+					.collect::<Vec<_>>(),
+				None,
+			),
+			FloodSettings::Sqrt { minimum } => (
+				self.context_data.peers
+					.iter()
+					.filter(|(_, peer)| peer.info.roles.is_full())
```
Comment (grbIzl) [inline client/network/src/protocol.rs:1247]: @arkpar was saying, that we have to include peers with authority role into the set for propagation anyway

**Comment by gnunicorn** [comment]
closing because of inactivity.

**Pull request closed without merging**